src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -1840,6 +1881,55 @@ public int ReceiveFrom(Span<byte> buffer, SocketFlags socketFlags, ref EndPoint
             return bytesTransferred;
         }
 
+        /// <summary>
+        /// Receives a datagram into the data buffer, using the specified <see cref="SocketFlags"/>, and stores the endpoint.
+        /// </summary>
+        /// <param name="buffer">A span of bytes that is the storage location for received data.</param>
+        /// <param name="socketFlags">A bitwise combination of the <see cref="SocketFlags"/> values.</param>
+        /// <param name="receivedSocketAddress">An <see cref="SocketAddress"/>, that will be updated with value of the remote peer.</param>
+        /// <returns>The number of bytes received.</returns>
+        /// <exception cref="ArgumentNullException"><c>remoteEP</c> is <see langword="null" />.</exception>
+        /// <exception cref="SocketException">An error occurred when attempting to access the socket.</exception>
+        /// <exception cref="ObjectDisposedException">The <see cref="Socket"/> has been closed.</exception>
+        public int ReceiveFrom(Span<byte> buffer, SocketFlags socketFlags, SocketAddress receivedSocketAddress)
+        {
+            ThrowIfDisposed();
+
+            ValidateBlockingMode();
+
+            int bytesTransferred;
+            SocketError errorCode = SocketPal.ReceiveFrom(_handle, buffer, socketFlags, receivedSocketAddress.Buffer, out int socketAddressSize, out bytesTransferred);
+
+            UpdateReceiveSocketErrorForDisposed(ref errorCode, bytesTransferred);
+            // If the native call fails we'll throw a SocketException.
+            SocketException? socketException = null;
+            if (errorCode != SocketError.Success)
+            {
+                socketException = new SocketException((int)errorCode);
+                UpdateStatusAfterSocketError(socketException);
+                if (NetEventSource.Log.IsEnabled()) NetEventSource.Error(this, socketException);
+
+                if (socketException.SocketErrorCode != SocketError.MessageSize)
+                {
+                    throw socketException;
+                }
+            }
+            else if (SocketsTelemetry.Log.IsEnabled())
+            {
+                SocketsTelemetry.Log.BytesReceived(bytesTransferred);
+                if (SocketType == SocketType.Dgram) SocketsTelemetry.Log.DatagramReceived();
+            }
+
+            if (socketException != null)
+            {
+                throw socketException;
+            }
```

Comment:
I'm confused by the structure of this code.  The only way socketException will be non-null is if we took the `if` branch above.  That branch only throws the socketException if it contains the right error code, but then this code immediately after it throws it unconditionally.  So should this just be deleted and the condition for the throw in the if branch be removed?

Reply:
I was wondering about that as well. I don't know what is expectation when we receive more than provided buffer. 

https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.receivefrom?view=net-7.0#system-net-sockets-socket-receivefrom(system-byte()-system-net-endpoint@)

The documentation suggest we would throw - but as you suggested there is no point of making is special case. 
I should perhaps move the size update up (or duplicate) so it would be also correct for partial receives.

Reply:
updated based on my best guess. We seems to lack test coverage for partial receives but I don't really want to drag it here.